File: leetcode/problems/cpp/two-sum.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "vector"
#include "unordered_map"
using namespace std;

class Solution {
public:
   vector<int> twoSum(vector<int>& nums, int target) {
      unordered_map <int, int> map;
      int i = 0;
      for (auto num : nums) {
         int t = target - num;
         auto searched = map.find(t);
         if (searched != map.end())
            return  { (*searched).second , i };
         map[nums[i]] = i;
         i++;
      }
      return { };
   }
};

File: leetcode/problems/rs/two-sum.rs
pub struct Solution;

use std::collections::HashMap;


impl Solution {
   pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
      let mut map = HashMap::new();

      for (i, num) in nums.into_iter().enumerate() {
         let diff = target - num;

         if let Some(&j) = map.get(&diff) {
            return vec![i as i32, j as i32];
         } else {
            map.insert(num, i);
         }
      }

      unreachable!()
   }
}
